File: android-tv-client/app/src/main/java/com/momo/ktv/tv/data/WebSocketClient.kt
package com.momo.ktv.tv.data

import android.util.Log
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import java.util.UUID
import java.util.concurrent.TimeUnit

class WebSocketClient(
    private val apiClient: ApiClient,
    private val scope: CoroutineScope
) {
    companion object {
        private const val TAG = "WSClient"
    }

    private val gson = Gson()
    private val client = OkHttpClient.Builder()
        .pingInterval(30, TimeUnit.SECONDS)
        .build()
    private var webSocket: WebSocket? = null
    private var reconnectJob: Job? = null

    val clientId = UUID.randomUUID().toString()
    val deviceId = "android-tv-${UUID.randomUUID().toString().take(8)}"

    // 回调
    var onQueueUpdate: ((List<QueueItem>) -> Unit)? = null
    var onControl: ((String, Map<String, Any>) -> Unit)? = null
    var onConnected: (() -> Unit)? = null
    var onDisconnected: (() -> Unit)? = null

    fun connect() {
        val url = apiClient.wsURL()
        Log.d(TAG, "Connecting to $url")
        try {
            val request = Request.Builder().url(url).build()
            webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(ws: WebSocket, response: Response) {
                Log.d(TAG, "WebSocket connected")
                onConnected?.invoke()
                sendRoleAnnounce()
            }

            override fun onMessage(ws: WebSocket, text: String) {
                handleMessage(text)
            }

            override fun onMessage(ws: WebSocket, bytes: ByteString) {
                handleMessage(bytes.utf8())
            }

            override fun onClosing(ws: WebSocket, code: Int, reason: String) {
                Log.d(TAG, "WebSocket closing: $code $reason")
            }

            override fun onClosed(ws: WebSocket, code: Int, reason: String) {
                Log.d(TAG, "WebSocket closed: $code $reason")
                onDisconnected?.invoke()
                scheduleReconnect()
            }

            override fun onFailure(ws: WebSocket, t: Throwable, response: Response?) {
                Log.e(TAG, "WebSocket failure: ${t.message}")
                onDisconnected?.invoke()
                scheduleReconnect()
            }
        })
        } catch (e: Exception) {
            Log.e(TAG, "WebSocket connect error: ${e.message}")
            onDisconnected?.invoke()
            scheduleReconnect()
        }
    }

    fun disconnect() {
        reconnectJob?.cancel()
        webSocket?.close(1000, "client disconnect")
        webSocket = null
    }

    private fun scheduleReconnect() {
        reconnectJob?.cancel()
        reconnectJob = scope.launch(Dispatchers.IO) {
            delay(3000)
            connect()
        }
    }

    private fun handleMessage(text: String) {
        try {
            val type = object : TypeToken<Map<String, Any>>() {}.type
            val json = gson.fromJson<Map<String, Any>>(text, type) ?: return
            val msgType = json["type"] as? String ?: return

            when (msgType) {
                "queue" -> {
                    val data = json["data"]
                    if (data != null) {
                        val dataJson = gson.toJson(data)
                        val qType = object : TypeToken<List<QueueItem>>() {}.type
                        val queue = gson.fromJson<List<QueueItem>>(dataJson, qType) ?: emptyList()
                        onQueueUpdate?.invoke(queue)
                    }
                }
                "control" -> {
                    val action = json["action"] as? String ?: return
                    val msgClientId = json["clientId"] as? String
                    if (msgClientId == clientId) return // 过滤自己的消息
                    val payload = json.filterKeys { it !in listOf("type", "action", "clientId") }
                    onControl?.invoke(action, payload)
                }
                "atmosphere", "blessing", "lyrics_updated", "lyrics_style" -> {
                    // 可扩展
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "handleMessage error: ${e.message}")
        }
    }

    private fun sendRoleAnnounce() {
        val msg = mapOf(
            "type" to "role_announce",
            "deviceId" to deviceId,
            "deviceName" to "Android TV",
            "role" to "player"
        )
        send(gson.toJson(msg))
    }

    fun sendControl(action: String, payload: Map<String, Any> = emptyMap()) {
        val msg: MutableMap<String, Any> = mutableMapOf(
            "type" to "control",
            "action" to action,
            "clientId" to clientId
        )
        msg.putAll(payload)
        send(gson.toJson(msg))
    }

    fun sendProgress(queueId: Int?, currentTime: Double, paused: Boolean, voice: String) {
        val msg = mutableMapOf(
            "type" to "progress",
            "deviceId" to deviceId,
            "currentTime" to currentTime,
            "paused" to paused,
            "voice" to voice
        )
        if (queueId != null) msg["queueId"] = queueId
        send(gson.toJson(msg))
    }

    fun sendPlaybackState(paused: Boolean, voice: String) {
        val msg = mapOf(
            "type" to "state",
            "paused" to paused,
            "voice" to voice
        )
        send(gson.toJson(msg))
    }

    private fun send(text: String) {
        webSocket?.send(text)
    }
}

File: ios-client/MomoKtviOS/WebSocketClient.swift
import Foundation
#if canImport(Starscream)
import Starscream
#endif

/// WebSocket 客户端：使用 Starscream 兼容 iOS 12+
/// 消息类型：control / queue / state / progress / role_announce / player_changed
class KTVWebSocketClient {
    static let shared = KTVWebSocketClient()

    #if canImport(Starscream)
    private var socket: WebSocket?
    #endif
    private var isConnected = false
    private var reconnectTimer: Timer?
    private var apiClient: KTVAPIClient?

    // 回调
    var onQueueUpdate: (([QueueItem]) -> Void)?
    var onControl: ((String, [String: Any]) -> Void)?
    var onConnected: (() -> Void)?
    var onDisconnected: (() -> Void)?

    private init() {}

    func connect(apiClient: KTVAPIClient) {
        self.apiClient = apiClient
        guard let wsURL = apiClient.wsURL() else { return }
        disconnect()

        #if canImport(Starscream)
        var request = URLRequest(url: wsURL)
        request.timeoutInterval = 10
        socket = WebSocket(request: request)
        socket?.delegate = self
        socket?.connect()
        #else
        // Fallback: 轮询模式（无 Starscream 时）
        startPolling()
        #endif
    }

    func disconnect() {
        reconnectTimer?.invalidate()
        reconnectTimer = nil
        #if canImport(Starscream)
        socket?.disconnect()
        socket = nil
        #endif
        isConnected = false
    }

    // MARK: - 发送消息
    func sendRoleAnnounce() {
        let msg: [String: Any] = [
            "type": "role_announce",
            "deviceId": apiClient?.deviceId ?? "",
            "deviceName": "iOS",
            "role": "player"
        ]
        send(msg)
    }

    func sendControl(_ action: String, payload: [String: Any] = [:]) {
        var msg: [String: Any] = [
            "type": "control",
            "action": action,
            "clientId": apiClient?.clientId ?? ""
        ]
        msg.merge(payload) { _, new in new }
        send(msg)
    }

    func sendProgress(queueId: Int?, currentTime: Double, paused: Bool, voice: String) {
        var msg: [String: Any] = [
            "type": "progress",
            "deviceId": apiClient?.deviceId ?? "",
            "currentTime": currentTime,
            "paused": paused,
            "voice": voice
        ]
        if let qid = queueId { msg["queueId"] = qid }
        send(msg)
    }

    func sendPlaybackState(paused: Bool, voice: String) {
        let msg: [String: Any] = [
            "type": "state",
            "paused": paused,
            "voice": voice
        ]
        send(msg)
    }

    private func send(_ msg: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: msg),
              let text = String(data: data, encoding: .utf8) else { return }
        #if canImport(Starscream)
        socket?.write(string: text)
        #endif
    }

    // MARK: - 消息处理
    private func handleMessage(_ text: String) {
        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = json["type"] as? String else { return }

        switch type {
        case "queue":
            if let qData = try? JSONSerialization.data(withJSONObject: json["data"] as Any),
               let queue = try? JSONDecoder().decode([QueueItem].self, from: qData) {
                DispatchQueue.main.async { self.onQueueUpdate?(queue) }
            }
        case "control":
            guard let action = json["action"] as? String else { return }
            let msgClientId = json["clientId"] as? String
            guard msgClientId != apiClient?.clientId else { return }
            let payload = json.filter { $0.key != "type" && $0.key != "action" && $0.key != "clientId" }
            DispatchQueue.main.async { self.onControl?(action, payload) }
        default:
            break
        }
    }

    // MARK: - 轮询降级（无 Starscream 时）
    private func startPolling() {
        reconnectTimer?.invalidate()
        reconnectTimer = Timer.scheduledTimer(withTimeInterval: 3.0, repeats: true) { [weak self] _ in
            self?.apiClient?.fetchQueue { queue in
                DispatchQueue.main.async { self?.onQueueUpdate?(queue) }
            }
        }
        onConnected?()
    }

    private func scheduleReconnect() {
        reconnectTimer?.invalidate()
        reconnectTimer = Timer.scheduledTimer(withTimeInterval: 3.0, repeats: false) { [weak self] _ in
            if let api = self?.apiClient { self?.connect(apiClient: api) }
        }
    }
}

#if canImport(Starscream)
extension KTVWebSocketClient: WebSocketDelegate {
    func didReceive(event: WebSocketEvent, client: WebSocketClient) {
        switch event {
        case .connected(let headers):
            isConnected = true
            DispatchQueue.main.async { self.onConnected?() }
            sendRoleAnnounce()
        case .disconnected(let reason, let code):
            isConnected = false
            DispatchQueue.main.async { self.onDisconnected?() }
            print("[WS] disconnected: \(code) \(reason)")
            scheduleReconnect()
        case .text(let text):
            handleMessage(text)
        case .binary(let data):
            if let text = String(data: data, encoding: .utf8) { handleMessage(text) }
        case .error(let error):
            isConnected = false
            print("[WS] error: \(error.localizedDescription)")
            DispatchQueue.main.async { self.onDisconnected?() }
            scheduleReconnect()
        case .cancelled:
            isConnected = false
        default:
            break
        }
    }
}
#endif
